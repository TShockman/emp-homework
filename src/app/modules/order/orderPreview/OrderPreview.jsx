import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import style from './orderPreview.scss';

export default class OrderPreview extends PureComponent {
  static propTypes = {
    order: PropTypes.shape({
      id: PropTypes.number.isRequired,
      ref: PropTypes.string.isRequired,
      status: PropTypes.string.isRequired,
      tracking: PropTypes.shape({
        carrier: PropTypes.string,
        trackingCode: PropTypes.string,
        status: PropTypes.string
      }),
      items: PropTypes.arrayOf(PropTypes.shape({
        sku: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
        amount: PropTypes.number.isRequired
      })).isRequired,
      discounts: PropTypes.arrayOf(PropTypes.shape({
        name: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired,
        value: PropTypes.number.isRequired
      }))
    }).isRequired
  };

  render() {
    const {order} = this.props;

    let tracking;
    if (order.tracking) {
      tracking = (
          <ul className="list-group">
            <li className="list-group-item">Carrier: <span className="fr">{order.tracking.carrier}</span></li>
            <li className="list-group-item">Tracking Code: <span className="fr">{order.tracking.trackingCode}</span></li>
            <li className="list-group-item">Status: <span className="fr">{order.tracking.status}</span></li>
          </ul>);
    } else {
      tracking = (
          <ul className="list-group">
            <li className="list-group-item">Not Available</li>
          </ul>);
    }

    const items = (
        <ul className="list-group">
          {order.items.map(item => (
            <li className="list-group-item">
              <span>{item.name} <small>({item.sku})</small> <b className="fr">${item.amount}</b></span>
            </li>
          ))}
        </ul>);

    let discounts = null;
    if (order.discounts) {
      discounts = (
          <div>
            <h5>Discounts:</h5>
            <div>
              <ul className="list-group">
                {order.discounts.map(discount => {
                  let discountAmt;
                  switch (discount.type) {
                    case 'percent':
                      discountAmt = <b className="fr">(-{discount.value}%)</b>;
                      break;
                    case 'amount':
                      discountAmt = <b className="fr">(-${discount.value})</b>;
                      break;
                    default:
                      discountAmt = null;
                  }
                  return (
                    <li className="list-group-item">
                      <span>{discount.name} {discountAmt}</span>
                    </li>
                  );
                })}
              </ul>
            </div>
          </div>
      );
    }

    return (
        <div className={cx("card", "orderPreview")}>
          <div className="card-header">
            <h3 className="card-title">Order {order.ref}</h3>
            <h4 className="card-subtitle">Status: {order.status}</h4>
          </div>
          <div className="card-body">
            <div>
              <h5>Items:</h5>
              <div>
                {items}
              </div>
            </div>
            {discounts}
            <div>
              <h5>Tracking:</h5>
              <div>
                {tracking}
              </div>
            </div>
            <button>View/Edit Order</button>
          </div>
        </div>
        );
  }
}