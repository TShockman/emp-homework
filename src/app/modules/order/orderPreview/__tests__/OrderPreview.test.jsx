import React from 'react';
import OrderPreview from '../';
import ReactTestRenderer from 'react-test-renderer/shallow';

const renderer = new ReactTestRenderer();

describe('OrderPreview', () => {
  it('Renders consistently with full information', () => {
    const props = {
      order: {
        id: 123,
        ref: "mockref-asdf1245",
        status: 'OK',
        tracking: {
          carrier: 'USPS',
          trackingCode: '123412353453254324',
          status: 'IN_TRANSIT'
        },
        items: [
          {
            sku: '123-123-123',
            name: 'Test Product',
            amount: 115
          },
          {
            sku: '456-456-456',
            name: 'Test Product Two',
            amount: 79
          }
        ],
        discounts: [
          {
            name: "Mock Family & Friends",
            type: 'percentage',
            value: 10
          },
          {
            name: "Christmas",
            type: 'amount',
            value: 10
          }
        ]
      }
    };
    expect(renderer.render(<OrderPreview {...props}/>)).toMatchSnapshot();
  });
});
