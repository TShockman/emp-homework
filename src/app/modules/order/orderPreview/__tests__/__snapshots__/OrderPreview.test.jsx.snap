// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderPreview Renders consistently with full information 1`] = `
<div
  className="card orderPreview"
>
  <div
    className="card-header"
  >
    <h3
      className="card-title"
    >
      Order 
      mockref-asdf1245
    </h3>
    <h4
      className="card-subtitle"
    >
      Status: 
      OK
    </h4>
  </div>
  <div
    className="card-body"
  >
    <div>
      <h5>
        Items:
      </h5>
      <div>
        <ul
          className="list-group"
        >
          <li
            className="list-group-item"
          >
            <span>
              Test Product
               
              <small>
                (
                123-123-123
                )
              </small>
               
              <b
                className="fr"
              >
                $
                115
              </b>
            </span>
          </li>
          <li
            className="list-group-item"
          >
            <span>
              Test Product Two
               
              <small>
                (
                456-456-456
                )
              </small>
               
              <b
                className="fr"
              >
                $
                79
              </b>
            </span>
          </li>
        </ul>
      </div>
    </div>
    <div>
      <h5>
        Discounts:
      </h5>
      <div>
        <ul
          className="list-group"
        >
          <li
            className="list-group-item"
          >
            <span>
              Mock Family & Friends
               
            </span>
          </li>
          <li
            className="list-group-item"
          >
            <span>
              Christmas
               
              <b
                className="fr"
              >
                (-$
                10
                )
              </b>
            </span>
          </li>
        </ul>
      </div>
    </div>
    <div>
      <h5>
        Tracking:
      </h5>
      <div>
        <ul
          className="list-group"
        >
          <li
            className="list-group-item"
          >
            Carrier: 
            <span
              className="fr"
            >
              USPS
            </span>
          </li>
          <li
            className="list-group-item"
          >
            Tracking Code: 
            <span
              className="fr"
            >
              123412353453254324
            </span>
          </li>
          <li
            className="list-group-item"
          >
            Status: 
            <span
              className="fr"
            >
              IN_TRANSIT
            </span>
          </li>
        </ul>
      </div>
    </div>
    <button>
      View/Edit Order
    </button>
  </div>
</div>
`;
